### DISCUSSION TITLE
HiLo IDs with dynamic prefix. What's the safest approach?

### DISCUSSION THREAD

**Discussion by mcrio** [ravendb/ravendb#15047] [Q&A] [answered]
Hi there, let's assume:

- We have a `Product` document
- We want to use HiLo for ID generation
- The prefix is dynamically constructed to include more data. Like: `products/tenant123/`
- End-goal: `products/tenant123/1-A`

I know there is an option to manually request the next HiLo number. What's the best option to generate a cluster unique Id in this case? We would request the next HiLo number and combine it manually with the node's tag?

Not sure why `GenerateNextIdForAsync` doesn't return the whole `1-A` identifier instead of just the number `1`. I guess there's some use case :)

p.s. I'm trying to avoid server-side generated IDs.

**Comment by ayende** [comment, marked as answer]
You need to call to `GenerateDocumentIdAsync` to get the full document id. 
Note that you have: `GenerateNextIdForAsync ` to allow you to do more interesting bits.

However...

HiLo assumes that you are going to use the "namespace" for a while, it takes _chunks_ out of the namespace. If you are going to do that on random locations, better to use an identity *or*, use a common namespace and update it.

So do something like: `"MyHilo"` prefix and then use `(await hilo.GenerateNextIdForAsync().Replace("MyHilo", "products/tenants123")`

**Reply by mcrio** [reply]
Hi Oren,

I don't mind sharing the HiLo between "tenants" (at least I don't see a problem at this point):
- products/tenant111/1-A
- products/tenant555/2-A
- products/tenant111/3-A

> Note that you have: GenerateNextIdForAsync to allow you to do more interesting bits.

Ok. This one allows me to get the next HiLo numeric value, I'd still need to append the node name to make the Id cluster-unique?

> So do something like: "MyHilo" prefix and then use (await hilo.GenerateNextIdForAsync().Replace("MyHilo", "products/tenants123")

I assume you meant `hilo.GenerateDocumentIdAsync` as `GenerateNextIdForAsync` returns just the HiLo number.

> You need to call to GenerateDocumentIdAsync to get the full document id.

It's just that this method is not part of `IHiLoIdGenerator` which is the type of `IDocumentStore.HiLoIdGenerator`.
To use `GenerateDocumentIdAsync` it seems I need to dig more into RavenDb internals which might lead to unforeseen problems as I don't know how it all works internally.

Two ways to use `GenerateDocumentIdAsync` I see:

```
// 1)
if (store.HiLoIdGenerator is AsyncMultiDatabaseHiLoIdGenerator asyncHiLoIdGenerator)
{
    string fullId  = await asyncHiLoIdGenerator.GenerateDocumentIdAsync(null, product);
}

// 2)
var multiDbHiLo = new AsyncMultiDatabaseHiLoIdGenerator(store as DocumentStore);
string fullId = await multiDbHiLo.GenerateDocumentIdAsync(null, product);
```

For option 2) I would need to reuse the generator throughout the app otherwise it'll get a new range each time it's instantiated.

**Summary**: I guess it could be great if `GenerateDocumentIdAsync` could be part of `IHiLoIdGenerator` that way we could generate the full id and replace the prefix as you suggested, without needing to play with the internals.

**Reply by ayende** [reply]
`GenerateDocumentIdAsync` will give you both the id and the tag `1-A`, frex.
The `IHiLoIdGenerator` is meant for giving you numeric id for when you need it, yes. I'm afraid that the string version isn't exposed, that is something that we can fix
You need to cast it to `AsyncMultiDatabaseHiLoIdGenerator`, yes. You should use your first option, yes.

**Reply by ayende** [reply]
https://issues.hibernatingrhinos.com/issue/RavenDB-19418/Expose-GenerateDocumentIdAsync-in-IHiloGenerator

**Reply by mcrio** [reply]
Hi Oren,

In case I just want to add a prefix to a HiLo auto generated ID, like prefixing the tenant id, is there any way to do that with Conventions?

If not, I guess I will need to use `GenerateDocumentIdAsync` once it will be exposed in:
https://issues.hibernatingrhinos.com/issue/RavenDB-19418/Expose-GenerateDocumentIdAsync-in-IHiloGenerator

**Reply by DejanMilicic** [reply]
Here is an example of custom HiLo generator
https://github.com/DejanMilicic/RavenWorkshop/blob/main/Northwind/Northwind/Features/Identifiers/CustomHiLoGenerator.cs

**Comment by mcrio** [comment]
Thanks @DejanMilicic 

I think I'll go with the following approach as it feels a bit less scary compared to extending current HiLo implementation classes :)

```
if (store.HiLoIdGenerator is not AsyncMultiDatabaseHiLoIdGenerator asyncHiLoIdGenerator)
{
    throw new Exception("Unexpected, check what changed...");
}
var book = new Book(){...};
var hiloId  = await asyncHiLoIdGenerator.GenerateDocumentIdAsync(store.Database, book);
book.Id = $"{book.TenantId}/{hiloId}";
await session.StoreAsync(book);
```

I thought maybe that would be possible using conventions but the problem is when using the `AsyncDocumentIdGenerator` convention, the default value is null until the store is initialized.

p.s. thanks for the repo link, has a lot of useful examples 👍
